### DISCUSSION TITLE
Suggestion: Selecting a tab from a setlist

### DISCUSSION THREAD

**Discussion by SuperSalmiakki** [sumatrapdfreader/sumatrapdf#4435] [Ideas]
I've been looking for a simple PDF Viewer that I could use on a Windows laptop with *touch interface* (portrait mode) during a musical performance to view the lyrics and Sumatra has worked pretty well (I like the fact that it remembers the _zooming_ and _last position_ of the viewed PDF). However, to make it better for at least that usage case, I am suggesting a couple of ideas for improvement.
**The problem**: When you have a set of song lyrics that is above a certain number (>8 or so) selecting the next song from the tab bar is nearly impossible (you accidentally hit the X and close the page) to navigate and select the song lyrics.
**Solutions** (_Full Screen mode_ - best for viewing lyrics, whole screen is available): One way would be to pick a _hot corner_ (user selectable?) that you could touch, and the UI would turn into a list of open tabs. Select the tab to move to by touching it. Fairly simple and I would think not hard to implement. Another way might be to _treat each tab as a page_, and you could move from one tab to another by swiping left/right (two fingers or tapping on a special area with your finger). There could be other (better) ways, but these come to my mind.
Related to this, as an extra cool feature, Sumatra could offer a way for the user to collect all songs in tabs (with proper placing and zooming for best visibility) and then save the whole set of tabs as a **set-list** that you could open and replace the currently opened tabs. Thus, the user could have multiple setlists ready to go and pick one and off you go!
Sumatra could also become a go to PDF viewer for lyrics if done right. It already is pretty good for a small number of tabs.

**Comment by GitHubRulesOK** [comment]
Ok most of that should be possible one way or another in current pre-release
however touch is not as versatile as programable pedal or foot keyboard.

Lets see there are several requests so I will start with the last to gather a collection of tabs need them to be set as required for store and open as a set list.

I have 8 tabs open full screen and a keyboard to save them from a preset personal command "another app all current tabs" call it what you want I dont need the menu as you can see I chose to use the `b` key for initiation.
![image](https://github.com/user-attachments/assets/26d2497c-c759-496c-96cd-47c9cc3e5ff2)

I will spare you the full details but it runs an external script so starts like this
![image](https://github.com/user-attachments/assets/95805089-6918-4513-b8e2-67c91a59c012)
and the result is another CMD file as seen here
![image](https://github.com/user-attachments/assets/cbc584d6-ec6a-4864-b673-47b707e43f11) which can be ANY contents but importantly for you has all the current tabbed files as a list for any app (including SumatraPDF) to run at once.
![image](https://github.com/user-attachments/assets/a81b7973-28d6-4824-a8db-f5b3b668224d)

So it is best to use a keyboard for saving pre-set sessions that can be run as a COMMAND selection of sets of any name and any number of files. To open fullscreen would need a tweak of that cmd file to include -fullscreen before the first filename.

**Reply by SuperSalmiakki** [reply]
Interesting! Creating and saving "setlist" or "sessions" by using a keyboard & mouse would be just fine. Setting up a song/lyrics list is anyways a separate action that you do beforehand. When performing, opening a session (a set of tabs with lyrics) in full screen mode and no keyboard (only touch available) should then open the PDFs as they were optimally setup for full screen viewing.

**Comment by GitHubRulesOK** [comment]
For other requests the use of tabbing has been significantly improved over the past week or so and can be used many ways to slide from one tab to the next via a single selectable key. so you need to look at pre-release for that too. and internal tab list is also improved. so that CTRL K can be assigned to a key also?

![image](https://github.com/user-attachments/assets/9c521ec1-0f1b-426e-ba78-5c7b0bdb6d15)

**Reply by SuperSalmiakki** [reply]
Thanks for all your efforts! I have to check the pre-release then. Just want to point out that, while performing, the 360 laptop is in portrait mode and only touch is available (laptop is like a tablet on a special music stand) so, while the setup can be done using keyboard & mouse, while performing all I have is a finger and a touch screen (simple and does not mess with the performance much). I would think that would be the case with most performers.

**Reply by GitHubRulesOK** [reply]
The limiting factor of touch only.
Is where there could be a problem as SumatraPDF is heavily into key based shortcuts and in full screen offers few options.

So firstly how would you be able to trigger a console or file command with portrait on one screen ?
If you use long touch do you see a menu? At best I think it will only be similar to this with only a list of single files
![image](https://github.com/user-attachments/assets/a4175c11-acf3-4929-9773-a589944f088e)
![image](https://github.com/user-attachments/assets/3ea70497-8810-48f3-ba32-b1ce6b418cbc)

you can preset sessions to offer different lists of favorites but that is several long / double presses between files for menu but not tabs 
Are there any button / programmable functions / gestures easily available so there is a good market of foot operated mini keyboard/programmable mice for musicians

**Reply by SuperSalmiakki** [reply]
Sorry, bear with me! I am not an expert in programming so take my ideas with a grain of salt but isn't the basic problem with how to initiate a pop-up window (with a simple list of names of the tabs) with full screen touch screen only? Once the selection from the list is done, switch to the selected tab should be just regular app operation even in touch only environment? The same can happen when in non-touch mode, too, in which case you would use a mouse/keys to select from the list (no need to press even OK/Cancel - you can always pop-up the list again). In non-touch mode the user can press Cntr-T (or something) to pop-up the list and make the selection. In touch mode the question then is how to turn a (single click/double click/long press) press on certain screen location (hot corner?) into Cntr-T? If the app can receive screen key presses, then checking its location on the screen should be fairly simple - right? Anyways, perhaps the challenge is that the app code should not have anything touch specific for portability's sake => Adding touch events would make it hard. Maybe this is just something that cannot be done easily (I totally understand if that is the case).
Just another idea: Do touch gestures generate certain key events? Like if you swipe up or down that typically means page up/down => They are key commands, too, to move page up & down. Perhaps swipe left/right (two fingers?) generate certain key commands that could be captured and turned into next/previous tab?

**Reply by GitHubRulesOK** [reply]
Especially for you :-) 
Pre-release today includes more touchy feely options BUT they are in the toolbar
For example to switch between many tabs you use maximal screen but with top menu visible

Ignore its Maths not music but see the Today toolbar has a LOOK button (any name you want) it is there by add in settings
```
Shortcuts [
	[
		Cmd = CmdCommandPalette @
		Key = @
		ToolbarText = LꙬK
	]
]
```

![image](https://github.com/user-attachments/assets/d77dfa09-777a-43f0-bf9e-0d7082fb08f3)

when you touch LooK it shows the vertical tabs so depends on finger nail / plectrum to choose the correct tab

![image](https://github.com/user-attachments/assets/05fa0979-0146-48e7-babe-a16815746ee1)

**Reply by SuperSalmiakki** [reply]
Alright! Thanks! I feel privileged! :-) I tried it and it is already pretty good. The visible tabs and the toolbar are not that bad (take some space - my old eyes need big fonts, so more space is always better ;-) And clicking "Look" works! Selecting the next tab from the menu works well, too. Making the tab menu items a bit bigger for finger operation (perhaps a setting?) would make it even better (sorry, nit picking)
BTW, I noticed this (at the bottom): [https://www.xda-developers.com/how-customize-touchpad-gestures-windows-11/](url)
It seems that, in Windows 11 at least, the user can record their own key command that would be executed as part of the chosen gesture. I have not tried it yet but that might solve the problem with touch interface? All you would need are Sumatra key commands for previous/next tab and the user can record a gesture to generate those key commands. Then Sumatra would not have to care about touch at all. The same procedure could be used for "LOOK", too: Generate the key command for "LOOK" (menu) for a specific gesture and the menu would pop up.

**Reply by GitHubRulesOK** [reply]
You can bump up UI icons and font sizes to suit (as I do for older chubby stumps) but there is a delay in spotting those new text icons dont follow so there is a new issue to be resolved after today :-)
```
ToolbarSize = 40
TreeFontName = automatic
TreeFontSize = 30
UIFontSize = 30
```
Looks like this:-
![image](https://github.com/user-attachments/assets/14965cb7-4d08-499e-87ed-e2b9b48868e0)


Gestures are hardware interface so SumatraPDF does not control keys or mice or touchpad but responds to their inputs.

The conversion of gesture to key commands like up down sideways are built by the hardware drivers. So will need the driver emulates a key character and SumatraPDF responds to that key, according to your setting SumatraPDF commands to keyboard characters.

So I assigned Look and  @ to be the same action, but the key character can be any available or externally provided.

**Reply by SuperSalmiakki** [reply]
OK, did not know that! I tried the font sizes, and it looks way better. A couple of issues thou (in single-page touch mode): The "Look" button does not fit, I presume, on the screen (the last one that fits is "Find:" and its text entry box). There is no way I can "scroll" the toolbar to see anything else. Also, when not in touch mode (regular laptop mode) the word "Look" does not follow the new font size (is smaller, the same size as before). Is there a way to get rid of some of the toolbar icons (settings)? I personally do not need rotates, Page number or printer when performing. Or perhaps they can be rearranged (I do not mind if some of the icons I do not need are not visible). Anyways, thanks again! This is getting better! 
Btw, can that key command (@) be a Control-Something (Cntr-X)?

**Reply by GitHubRulesOK** [reply]
The reason there were very few additions to icons for a decade or so was the minimalist appeal of "all keys few icons"  but last addition was probably the 2 you dont want/need but considered "essential" by others.

Portrait netbook tablets like mine cant display more than the old ones, hence I tended to use 3rd party outside dropdown systems for more elaborate scripting, whilst others hijacked the few that were included but that requires much more complex hooking the screen locations or edit the source code. I had to fiddle with icon sizes to get the old sets to fit narrow screen width!

Currently existing icons cannot be discounted. You could raise that as a formal issue, but it may already have been raised so you should comment there rather than open a duplicate.

I have not looked recently at the old touch driver but recollect I could set compass point swiping to keys so at least 8 different ways ?

If you can invoke `Ctrl + X` by swipe/gesture then yes it may be possibly used to replace the @

**Reply by GitHubRulesOK** [reply]
>Also, when not in touch mode (regular laptop mode) the word "Look" does not follow the new font size (is smaller, the same size as before). 

That should be better in Tomorrow's update :-) so wait sleep download repeat https://github.com/sumatrapdfreader/sumatrapdf/issues/4443

**Reply by GitHubRulesOK** [reply]
@SuperSalmiakki 
That update is now available so no need to sleep just update however width seems to be odd ! YourMilageMayVary
![image](https://github.com/user-attachments/assets/c2cc0fde-9d37-422e-aaec-f3015e068f65)

**Reply by SuperSalmiakki** [reply]
I set the toolbar font size to 25 in order to get everything to fit. Here is how it looks like in tablet mode (finally learned how to take screen shots when in tablet mode):
![image](https://github.com/user-attachments/assets/fcddd927-46ba-458c-9980-472993ff4a2f)
For some reason there is unnecessary extra space after the page number and the "Look" button does not fit ("Lo...") even though there is space to the right of it for another full "Look" button. Hope this helps!
I also checked that **Ctrl + L** (for example) works with the keyboard AND even with _touch pad_ (you can specify what a 3-finger tap does and in this case, it generates Ctrl + L) BUT these gestures are only for the **touch pad** (not **touch screen** - at least I was not able to get it to work). Thus, I can open the tab list just by 3-finger tapping on the touch pad (when not in tablet mode).
Anyways, if you can get rid of the extra white space in the toolbar and let the "Look" button to take whatever space it needs at the end of the toolbar then I can try to find the right toolbar font size so that everything fits but no extra space is left (everything is as big as it can be without losing any of the toolbar items). Then I can just use the "Look" button to move to the next tab (PDF). 
Thank you, once again, for the work you do! I know you do not have to do any of this so greatly appreciated!

**Reply by GitHubRulesOK** [reply]
PS. there is yet another tweak today for tomorrow ?
https://github.com/sumatrapdfreader/sumatrapdf/commit/402c7695ca4d01541fccf027a145590e4e12b678

Well it is cutting edge days the big gap is for 4 fat ladies ie. 8888 pages it has taken a decade or so of constant changes to get to this static view :-) and yes the Lo.. can be bettered given a few more free hours. I am only the old wizened front house robot. The "breaking" default  is likely between 20-24 pixels as that's what the "square" ones were defaulted to 

![image](https://github.com/user-attachments/assets/2423180b-e93d-419a-826f-b42463d2aabc)

Hardware behaviour is a problem down to the driver abilities and whilst mice may have extra programmable buttons there is no requirement for touch to do much except pinch and punch it. I dont use mine enough to have problems apart from sizes, but have asked on occasion for "extras" or found 3rd party alternative for the required controls at the time.

**Reply by SuperSalmiakki** [reply]
I downloaded the latest pre-release like 6 hour ago and used that - seems to be same that is there now. And yeah, I am also only old front house performer :-)

**Reply by GitHubRulesOK** [reply]
I especially mentioned tomorrow as **usually** daily updates are about 3AM UCT ? and on occasion fail to be any for a week or more
but there were some pushed early manually yesterday so gives the impression of very early fixes (or fails) always back-up a reasonably good one.

Work in Progress 15664+1000 is 3.6.16664 last accessible daily is 16649 (the small change I listed untested by me, is between the 2 so not included yet, and may even be unnoticeable unless one icon tips onto page from half unseen to seen)

**Reply by GitHubRulesOK** [reply]
@SuperSalmiakki 
it looks like daily failed to launch so currently no change for above however I have raised an issue to ask for a fresh take on the size problem see https://github.com/sumatrapdfreader/sumatrapdf/issues/4449

**Reply by SuperSalmiakki** [reply]
FYI: I used Sumatra as part of one of my performances last weekend and it worked quite well even as it is (quite small button/text to select the tab menu). One problem thou: I thought that Sumatra would keep the visual settings (zoom and position) for each PDF when you reopen them, and it does most of the time. However, if I load a set of PDFs, set them up properly (zoom and position) and select each tab everything works as it should. I can even close Sumatra and reopen and it opens the tabs properly as they were setup - most of the time. Maybe time has something to do with it? If I open Sumatra later again some or all PDFs have lost their zoom & position => defaults. I could not figure out when or why it happens. Perhaps going in and out of tablet mode has something to do with it?

**Reply by GitHubRulesOK** [reply]
Hmm 
for one session each "file" setting is remembered at the time of closure (affected by layout), so I cannot guess your settings but say a music sheet has several instruments so you save page 6 zoomed on guitar part as single page.
```
FilePath = C:\Users\me\music.pdf
		Favorites [
		]
		IsPinned = false
		IsMissing = false
		OpenCount = 2
		UseDefaultState = false
		DisplayMode = continuous
		ScrollPos = -1 304.441
		PageNo = 6
		Zoom = fit width
		Rotation = 0
		WindowState = 1
		WindowPos = 59 144 1835 937
		ShowToc = false
		SidebarDx = 340
		DisplayR2L = false
		ReparseIdx = 0
```

That will be the intended start for next time opening, but if during the session you turn pages then on closure the page number will be now say 8 not 6 also the starting zoom next time may be different to when last saved. Thus you need to save all those sheets exactly as needed for any fresh start. BEWARE the data you see on opening settings MAY NOT be the current state, just the starting state, that may have altered in memory and thus differ on close/save. **The "remember" state should be saved on tab closure**

Without watching your movements I can only suggest test a few steps at a time so you understand where any variation happens.

A common problem could be 2 tabs from same PDF as the **"file"** states could conflict, so in such case "single" pages 1.pdf 2.pdf may be better. **HOWEVER tab states should work for each tab** so here page 6 and page 3 have different scroll positions
```
SessionData [
	[
		TabStates [
			[
				FilePath = C:\Users\my\music.pdf
				DisplayMode = continuous
				PageNo = 6
				Zoom = fit width
				Rotation = 0
				ScrollPos = -1 -1
				ShowToc = false
			]
			[
				FilePath = C:\Users\my\music.pdf
				DisplayMode = continuous
				PageNo = 3
				Zoom = fit width
				Rotation = 0
				ScrollPos = -1 304.232
				ShowToc = false
			]
		]
		TabIndex = 2
		WindowState = 1
		WindowPos = 809 54 989 937
		SidebarDx = 340
	]
]
```

**Reply by GitHubRulesOK** [reply]
@SuperSalmiakki 

OK thinking about your method/set-up I should perhaps have suggested a different starting point.

TABS are only retained by saving the **whole settings.txt** file at a given point in time. Thus my suggestion to save file states may not work as per your desire to save a session of "tabs". for that to work well you need to close session and save the settings text for future use, HOWEVER there can only be ONE settings file.

To start with different  settings.TXT there is only one switch **-appdata** which expects a "folder" name thus needs different starting folders with or without other files ! then you startup sequence would need to point to the different start folder.
`sumatrapdf.exe -appdata c:\mydata\weddings` where weddings folder has its own sumatrapdf-settings.txt

**Reply by SuperSalmiakki** [reply]
Thanks for the info! Just to clarify: At least in my case, each PDF is only one page (don't have time or possibility to scroll while performing) that is position and zoom so that there is no unnecessary white space around (max zoom without losing any of the content) => Each tab is a one-page PDF positioned & zoomed correctly. 
What you say about the settings file makes sense and I thought of copying and saving a session of tabs by copying the settings.txt file for later use (settings file for various song lists). You could then copy and replace the current settings file with the one you need. I guess another way would be to separate the app settings info (sumatra.ini) from the PDF setting info (pdf-settings.txt). PDF settings could then have the latest positioning & zoom settings for all PDFs that you have opened. The ini file can have the names of the last opened tabs or there could be a separate "session.txt" file to denote that. Anyways, that would require a fair amount of work so just ignore me.

**Reply by GitHubRulesOK** [reply]
it WAS (is?) possible to start sumatrapdf with a batch built limited settings file and sumatrapdf start by filling in the defaults HOWEVER that also has several challenges, so not a reasonable working solution for this case.

If you have pre-set settings.txt then your idea of copy over the "saved **session**" for each different launching is probably the simplest to manage. For minor adjustments you simply edit or resave each session so for example a cmd or shortcut icon file that simply does `copy c:\data\weddings.txt c:\path\folder\sumatrapdf-settings.txt & sumatrapdf.exe`

**Reply by Ronald-G-Andrade** [reply]
> ![image](https://private-user-images.githubusercontent.com/12754913/355109978-14965cb7-4d08-499e-87ed-e2b9b48868e0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4MTk3MjYsIm5iZiI6MTcyMzgxOTQyNiwicGF0aCI6Ii8xMjc1NDkxMy8zNTUxMDk5NzgtMTQ5NjVjYjctNGQwOC00OTllLTg3ZWQtZTJiOWI0ODg2OGUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE2VDE0NDM0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRhNzNlYzE5N2M2NDVhNTkwYmNlMjc4OWE2ZjNhMDViOTcxNzE0ZmRhMDMzMjMwNTVmNTdmMjY2YzQ4ZGIyN2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._fNjT3IqquujmCN6_eyvwa7ZDcqQnzovCVxxZFxaY0g)

Hello, thanks for this new feature, it's really nice. But please, is there any way it can be disabled?? Because I tend to compare two PDFs by pressing Tab key many times, while I keep pressing Ctrl key, but this is not possible with this new “Tab Search” menu. Thank you.

**Reply by GitHubRulesOK** [reply]
@Ronald-G-Andrade  wrong thread This one is about saving tabs. However the tab right left function was shifted from tabs to page "Same as browsers" so CTRL + page UP/DOWN 
For 2 tabs set one icon to go tab right and it will cycle 1 > 2 > 1 >  2

**Comment by GitHubRulesOK** [comment]
My touch tablet is very old and is only 180 :-)
however it has lots of bells and whistles like OSKD or is it OCD abilities but I would need to move task bars to top or bottom to run extra commands and maximise screen usage for reading.
![image](https://github.com/user-attachments/assets/55481a02-f3d1-4b22-bcad-fa6e19efbea4)
